backend/common/src/main/java/ai/verta/modeldb/common/db/migration/Migrator.java

```diff
@@ -22,19 +31,147 @@ public Migrator(Connection connection, String resourcesDirectory) {
   /** Assumes that the underlying database has already been created. */
   public void performMigration(RdbConfig config) throws SQLException, MigrationException {
     log.info("Starting database migration process");
-    MigrationDatastore migrationDatastore = MigrationDatastore.create(config, connection);
+    performMigration(config, null);
+  }
 
+  void performMigration(RdbConfig config, Integer desiredVersion)
+      throws SQLException, MigrationException {
+    MigrationDatastore migrationDatastore = setupDatastore(config);
+    MigrationState currentState = findCurrentState();
+
+    if (currentState.isDirty()) {
+      // todo: implement dirty handling
+      throw new MigrationException(
+          "Database is in a dirty state. Bailing out until dirty handling is implemented.");
+    }
+    int versionToTarget =
+        desiredVersion != null
+            ? desiredVersion
+            : gatherMigrations(migration -> true, migration -> true).stream()
+                .map(Migration::getNumber)
+                .max(Integer::compareTo)
+                .orElse(0);
+    log.info("Starting database migration process to version: " + versionToTarget);
+
+    if (currentState.getVersion() == versionToTarget) {
+      log.info("No migrations to perform. Database is already at version " + versionToTarget);
+      return;
+    }
+    NavigableSet<Migration> migrationsToPerform =
```

Comment:
TIL about `NavigableSet` :)

Reply:
I re-learned about it during this work, myself. I found it when I went to reverse the order of the TreeSet and found that it returned a NavigableSet!

Reply:
Are you using any of the methods of `NavigableSet`?  If not, let's use  `SortedSet` or even `Set`.  Unless it being a `NavigableSet` is meaningful (or is likely to be meaningful), we should ignore that detail.

Reply:
We are using the `descendingSet` method from NavigableSet when applying the `down` migrations.

Reply:
I wouldn't want to return a bare `Set` or declare this variable as one, since the ordering is critical to the in-order application of the migrations. I could see returning SortedSet, but that's also a pretty arcane interface in the JDK, so I don't know that it would be any clearer in this case. I don't feel strongly about that, though.

Reply:
apropos of nothing, I recently saw a JEP that proposes introducing [IIRC] `Traversable` as an interface that would encompass anything that allowed backward and forward in-order navigation of data in a collection.

Reply:
`SortedSet` is more immediately understandable to me that `NavigableSet`.